wasmCloud/wasmCloud
# feat: support OTEL traces end-to-end

## Feature or Problem
This PR:
- adds OTEL config values to the CLI args for the host
- sets up tracing for the host
- proxies OTEL config values to providers via `HostData`
- uses a strongly-typed `HostData` in the host to pass data to providers
- adds a `core` crate for types shared across the wasmCloud ecosystem (primarily between host and provider)
- adds a `tracing` crate for setting up tracing and extracting/injecting trace context on invocations (this code was migrated from the provider SDK)

## Related Issues
Resolves https://github.com/wasmCloud/wasmCloud/issues/568
Resolves https://github.com/wasmCloud/wasmCloud/issues/577
Resolves https://github.com/wasmCloud/wasmCloud/issues/563

## Release Information
Next

## Consumer Impact
Providers built against the updated provider SDK will need to run on a host that supplies `OtelConfig`

## Testing



Built on platform(s)


Tested on platform(s)

### Unit Test(s)


### Acceptance or Integration


### Manual Verification
I set up tempo locally along with the kvcounter actor + HTTP server and KV Redis provider built using these changes. Here is an example where you can see spans from the host and both providers as part of a round-trip invocation:

```
docker exec -it 8a6428c77728 /bin/sh -c "wget -q -O - localhost:8000/api/traces/cef9a8119b196d111bc2f5956f22cd4e" | jq -r '.batches | .[] | .instrumentationLibrarySpans | .[] | .spans | .[] | .name' | tr '\n' ' '

increment dispatch publish rpc instantiate instantiate instantiate write_bytes write_bytes guest_request read_bytes read_string read_bytes read_string read_bytes read_string read_bytes identify_wasmbus_target identify_wasmbus_target call_operation_with_payload call_sync handle set_host_response host_call host_response_len write_bytes host_response read_bytes guest_response call call handle handle_invocation handle_call request inner_rpc call_actor http request request 
```
